tests/src/InputCompletion.cpp

```diff
@@ -0,0 +1,336 @@
+#include "Application.hpp"
+#include "BaseSettings.hpp"
+#include "common/Aliases.hpp"
+#include "common/CompletionModel.hpp"
+#include "controllers/accounts/AccountController.hpp"
+#include "messages/Emote.hpp"
+#include "mocks/EmptyApplication.hpp"
+#include "mocks/Helix.hpp"
+#include "providers/twitch/TwitchIrcServer.hpp"
+#include "singletons/Emotes.hpp"
+#include "singletons/Paths.hpp"
+#include "singletons/Settings.hpp"
+#include "widgets/splits/InputCompletionPopup.hpp"
+
+#include <boost/optional/optional_io.hpp>
+#include <gtest/gtest.h>
+#include <QDir>
+#include <QFile>
+#include <QModelIndex>
+#include <QString>
+
+namespace {
+
+using namespace chatterino;
+using ::testing::Exactly;
+
+class MockTwitchIrcServer : public ITwitchIrcServer
+{
+public:
+    const BttvEmotes &getBttvEmotes() const override
+    {
+        return this->bttv;
+    }
+
+    const FfzEmotes &getFfzEmotes() const override
+    {
+        return this->ffz;
+    }
+
+    const SeventvEmotes &getSeventvEmotes() const override
+    {
+        return this->seventv;
+    }
+
+    BttvEmotes bttv;
+    FfzEmotes ffz;
+    SeventvEmotes seventv;
+};
+
+class MockApplication : EmptyApplication
+{
+public:
+    AccountController *getAccounts() override
+    {
+        return &this->accounts;
+    }
+
+    ITwitchIrcServer *getTwitch() override
+    {
+        return &this->twitch;
+    }
+
+    IEmotes *getEmotes() override
+    {
+        return &this->emotes;
+    }
+
+    AccountController accounts;
+    MockTwitchIrcServer twitch;
+    Emotes emotes;
+};
+
+}  // namespace
+
+namespace chatterino {
+
+class MockChannel : public Channel
+{
+public:
+    MockChannel(const QString &name)
+        : Channel(name, Channel::Type::Twitch)
+    {
+    }
+};
+
+}  // namespace chatterino
+
+EmotePtr namedEmote(const EmoteName &name)
+{
+    return std::shared_ptr<Emote>(new Emote{
+        .name{name},
+        .images{},
+        .tooltip{},
+        .zeroWidth{},
+        .id{},
+        .author{},
+    });
+}
+
+void addEmote(EmoteMap &map, const QString &name)
+{
+    EmoteName eName{.string{name}};
+    map.insert(std::pair<EmoteName, EmotePtr>(eName, namedEmote(eName)));
+}
+
+static QString DEFAULT_SETTINGS = R"!(
+{
+    "accounts": {
+        "uid117166826": {
+            "username": "testaccount_420",
+            "userID": "117166826",
+            "clientID": "abc",
+            "oauthToken": "def"
+        },
+        "current": "testaccount_420"
+    }
+})!";
+
+class InputCompletionTest : public ::testing::Test
+{
+protected:
+    void SetUp() override
+    {
+        // Write default settings to the mock settings json file
+        ASSERT_TRUE(QDir().mkpath("/tmp/c2-tests"));
```

Comment:
It would be awesome if the tests could run on all supported platforms. Consider using [`QTemporaryDir`](https://doc.qt.io/qt-6/qtemporarydir.html) to create a temporary directory.

Reply:
yes, but should not be done in this PR